### PR TITLE
Single execution of every method

### PR DESCRIPTION
**Pull request by JeffreyDevloo**

### CODE REVIEW

File: ovs/extensions/healthcheck/decorators.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                                raise TypeError('Expected an instance of {}'.format(type(HCResults.HCResultCollector)))
+                            kwargs['result_handler'] = result_handler
+                    return callback_func(*tuple(arguments), **kwargs)
+            finally:
+                _mutex.release()
+        return wrapped
+    return wrapper
+
+
+def ensure_single_with_callback_local(key, callback=None):
```
Comment (khenderick) [inline ovs/extensions/healthcheck/decorators.py]: Can't you merge these?

File: ovs/extensions/healthcheck/decorators.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+# as published by the Free Software Foundation, in version 3 as it comes
+# in the LICENSE.txt file of the Open vStorage OSE distribution.
+#
+# Open vStorage is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY of any kind.
+import inspect
+from functools import wraps
+from ovs.extensions.healthcheck.helpers.cache import CacheHelper
+from ovs.extensions.healthcheck.result import HCResults
+from ovs.extensions.generic.system import System
```
Comment (khenderick) [inline ovs/extensions/healthcheck/decorators.py]: This import seems out of order here.

File: ovs/extensions/healthcheck/decorators.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+                _mutex.acquire(wait=0.005)
+                local_sr = System.get_my_storagerouter()
+                CacheHelper.set(key=key, item={'ip': local_sr.ip, 'hostname': local_sr.name}, expire_time=60)
+                return func(*args, **kwargs)
+            except NoFileLockAvailableException:
+                if callback is None:
+                    return
+                else:
+                    callback_func = callback.__func__ if isinstance(callback, staticmethod) else callback
+                    argnames, _, _, _ = inspect.getargspec(callback_func)
```
Comment (khenderick) [inline ovs/extensions/healthcheck/decorators.py]: Can you use `[0]` here?

File: ovs/extensions/healthcheck/testing/decorators.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+import threading
+import unittest
+from ovs.extensions.healthcheck.decorators import ensure_single_with_callback_cluster
+from ovs.extensions.healthcheck.decorators import ensure_single_with_callback_local
+
+
+class ClusterCheckTester(unittest.TestCase):
+
+    def setUp(self):
+        self.concurreny_amount = 5
```
Comment (khenderick) [inline ovs/extensions/healthcheck/testing/decorators.py]: This looks like actual test logic inside the `setUp`. Wouldn't it make more sense to move this into the test itself?

File: ovs/extensions/healthcheck/testing/decorators.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            kwargs = {}
+        thread = threading.Thread(target=target, args=tuple(args), kwargs=kwargs)
+        thread.setName(str(name))
+        thread.start()
+        return thread
+
+    @staticmethod
+    def concurrency_callback(*args, **kwargs):
+        thread_object = kwargs.get('thread_object')
+        thread_object['callbacks'][time.time()] = 'Called'
```
Comment (khenderick) [inline ovs/extensions/healthcheck/testing/decorators.py]: Possible that the key is overwritten and the test would fail.

**Comment by JeffreyDevloo** [comment]
selection_119

**Pull request merged**